src/datachain/toolkit/split.py

```diff
@@ -56,16 +67,28 @@ def train_test_split(dc: DataChain, weights: list[float]) -> list[DataChain]:
     if any(weight < 0 for weight in weights):
         raise ValueError("Weights should be non-negative")
 
-    weights_normalized = [weight / sum(weights) for weight in weights]
+    def scale(multiplier: float) -> int:
+        if multiplier < 0:
+            multiplier = 0.0
+        if multiplier > 1:
+            multiplier = 1.0
+        numerator, denominator = float(multiplier).as_integer_ratio()
+        return MAX_SIGNED_INT64 * numerator // denominator
 
-    resolution = 2**31 - 1  # Maximum positive value for a 32-bit signed integer.
+    weights_normalized = [weight / sum(weights) for weight in weights]
+    limits = [
+        [
+            scale(sum(weights_normalized[:index])),
+            scale(sum(weights_normalized[: index + 1])),
+        ]
+        for index, _ in enumerate(weights_normalized)
+    ]
 
+    rand_col = "sys__rand" if seed is None else int_hash_64(bit_xor("sys__rand", seed))
```

Comment:
it would be great to know how beadly we need `int_hash_64`. I'd understand the concern if we would be XORing with `sys.id` which is not uniformally distributed. However, `sys.rand` should not have any skews.

Reply:
I understand `inthash64` looks like we don't really need it.

I was trying XORing `sys.rand` with `seed` and different big prime numbers in different combinations, but I can not make it works to get good enough distribution. In two words: using `seed` gives not much difference with default "no seed" split.

I was also trying to use `murmurhash`, `siphash` and other hash algorithms, as they are pretty common, implemented in ClickHouse and can be added to SQLite, but they requires installing additional packages which I was trying to avoid (and I can, but I don't really want to implement those by myself). Also they might not be available in other databases.

If anyone have suggestions on good algorithm using XOR only, I'm all in. I can also spent more time on this.

Reply:
Resolved with #678